Flacky test RoundRobinStrategyTest.testSkipConnections

Sometimes this test fails with 
```
java.util.concurrent.CompletionException: org.opentest4j.AssertionFailedError: Unexpected exception thrown: io.tarantool.driver.exceptions.NoAvailableConnectionsException: No available connections

	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:320)
	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:325)
	at java.base/java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1809)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
	at java.base/java.lang.Thread.run(Thread.java:1676)
Caused by: org.opentest4j.AssertionFailedError: Unexpected exception thrown: io.tarantool.driver.exceptions.NoAvailableConnectionsException: No available connections
	at org.junit.jupiter.api.AssertDoesNotThrow.createAssertionFailedError(AssertDoesNotThrow.java:83)
	at org.junit.jupiter.api.AssertDoesNotThrow.assertDoesNotThrow(AssertDoesNotThrow.java:54)
	at org.junit.jupiter.api.AssertDoesNotThrow.assertDoesNotThrow(AssertDoesNotThrow.java:37)
	at org.junit.jupiter.api.Assertions.assertDoesNotThrow(Assertions.java:3005)
	at io.tarantool.driver.core.RoundRobinStrategyTest.lambda$testSkipConnections$9(RoundRobinStrategyTest.java:119)
	at java.base/java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1806)
	... 3 more
Caused by: io.tarantool.driver.exceptions.NoAvailableConnectionsException: No available connections
	at io.tarantool.driver.core.TarantoolConnectionSelectionStrategies$RoundRobinStrategy.next(TarantoolConnectionSelectionStrategies.java:57)
	at io.tarantool.driver.core.RoundRobinStrategyTest.lambda$testSkipConnections$8(RoundRobinStrategyTest.java:119)
	at org.junit.jupiter.api.AssertDoesNotThrow.assertDoesNotThrow(AssertDoesNotThrow.java:50)
	... 7 more
```
and sometimes it does pass.

`@RepeatedTest(value = 100)` does not help in reproducing - it seems that only first repetition can fail and all the others will always pass.

--- a/src/main/java/io/tarantool/driver/api/TarantoolClient.java
+++ b/src/main/java/io/tarantool/driver/api/TarantoolClient.java
@@ -60,6 +60,8 @@ public interface TarantoolClient extends AutoCloseable, TarantoolCallOperations,
 
     /**
      * Get the default factory for {@link TarantoolTuple} instances
+     *
+     * @return tuple factory instance
      */
     TarantoolTupleFactory getTarantoolTupleFactory();
 }

--- a/src/main/java/io/tarantool/driver/api/conditions/Conditions.java
+++ b/src/main/java/io/tarantool/driver/api/conditions/Conditions.java
@@ -204,6 +204,7 @@ public final class Conditions {
      *
      * @param tuple last tuple value from the previous result, may be null
      * @param tupleConverter converter of the specified tuple type into a MessagePack array
+     * @param <T> tuple type
      * @return new {@link Conditions} instance
      */
     public static <T> Conditions after(@Nullable T tuple, ObjectConverter<T, ArrayValue> tupleConverter) {
@@ -227,6 +228,7 @@ public final class Conditions {
      *
      * @param tuple last tuple value from the previous result, may be null
      * @param tupleConverter converter of the specified tuple type into a MessagePack array
+     * @param <T> tuple type
      * @return new {@link Conditions} instance
      */
     public <T> Conditions startAfter(T tuple, ObjectConverter<T, ArrayValue> tupleConverter) {
@@ -964,7 +966,7 @@ public final class Conditions {
         private final T tuple;
         private final ObjectConverter<T, ArrayValue> tupleConverter;
 
-        public StartTupleWrapper(T tuple, ObjectConverter<T, ArrayValue> tupleConverter) {
+        StartTupleWrapper(T tuple, ObjectConverter<T, ArrayValue> tupleConverter) {
             this.tuple = tuple;
             this.tupleConverter = tupleConverter;
         }

--- a/src/main/java/io/tarantool/driver/api/space/ProxyTarantoolSpace.java
+++ b/src/main/java/io/tarantool/driver/api/space/ProxyTarantoolSpace.java
@@ -259,6 +259,11 @@ public class ProxyTarantoolSpace implements TarantoolSpaceOperations {
     }
 
     @Override
+    public TarantoolSpaceMetadata getMetadata() {
+        return spaceMetadata;
+    }
+
+    @Override
     public String toString() {
         return String.format("ProxyTarantoolSpace [%s]", spaceName);
     }

--- a/src/main/java/io/tarantool/driver/api/space/TarantoolSpace.java
+++ b/src/main/java/io/tarantool/driver/api/space/TarantoolSpace.java
@@ -301,6 +301,11 @@ public class TarantoolSpace implements TarantoolSpaceOperations {
     }
 
     @Override
+    public TarantoolSpaceMetadata getMetadata() {
+        return spaceMetadata;
+    }
+
+    @Override
     public String toString() {
         return String.format("TarantoolSpace %s [%d]", spaceMetadata.getSpaceName(), spaceMetadata.getSpaceId());
     }

--- a/src/main/java/io/tarantool/driver/api/space/TarantoolSpaceOperations.java
+++ b/src/main/java/io/tarantool/driver/api/space/TarantoolSpaceOperations.java
@@ -6,6 +6,7 @@ import io.tarantool.driver.api.tuple.TarantoolTuple;
 import io.tarantool.driver.exceptions.TarantoolClientException;
 import io.tarantool.driver.mappers.ValueConverter;
 import io.tarantool.driver.api.tuple.operations.TupleOperations;
+import io.tarantool.driver.metadata.TarantoolSpaceMetadata;
 import org.msgpack.value.ArrayValue;
 
 import java.util.concurrent.CompletableFuture;
@@ -184,4 +185,11 @@ public interface TarantoolSpaceOperations {
                                                      TarantoolTuple tuple,
                                                      TupleOperations operations,
                                                      ValueConverter<ArrayValue, T> tupleMapper);
+
+    /**
+     * Get metadata associated with this space
+     *
+     * @return space metadata
+     */
+    TarantoolSpaceMetadata getMetadata();
 }

--- a/src/main/java/io/tarantool/driver/api/tuple/TarantoolFieldImpl.java
+++ b/src/main/java/io/tarantool/driver/api/tuple/TarantoolFieldImpl.java
@@ -68,8 +68,12 @@ public class TarantoolFieldImpl implements TarantoolField {
 
     @Override
     public boolean equals(Object o) {
-        if (this == o) return true;
-        if (!(o instanceof TarantoolFieldImpl)) return false;
+        if (this == o) {
+            return true;
+        }
+        if (!(o instanceof TarantoolFieldImpl)) {
+            return false;
+        }
         TarantoolFieldImpl that = (TarantoolFieldImpl) o;
         return Objects.equals(value, that.value);
     }

--- a/src/main/java/io/tarantool/driver/api/tuple/TarantoolNullField.java
+++ b/src/main/java/io/tarantool/driver/api/tuple/TarantoolNullField.java
@@ -6,6 +6,8 @@ import org.msgpack.value.Value;
 import org.msgpack.value.ValueFactory;
 import org.springframework.lang.Nullable;
 
+import java.util.Objects;
+
 /**
  * Represents a field with empty value which serializes to {@code msgpack.NULL}
  *
@@ -26,7 +28,14 @@ public final class TarantoolNullField implements TarantoolField {
 
     @Override
     public boolean equals(Object o) {
-        if (this == o) return true;
+        if (this == o) {
+            return true;
+        }
         return o instanceof TarantoolNullField;
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hashCode(this);
     }
 }

--- a/src/main/java/io/tarantool/driver/api/tuple/TarantoolTupleImpl.java
+++ b/src/main/java/io/tarantool/driver/api/tuple/TarantoolTupleImpl.java
@@ -384,8 +384,12 @@ public class TarantoolTupleImpl implements TarantoolTuple {
 
     @Override
     public boolean equals(Object o) {
-        if (this == o) return true;
-        if (!(o instanceof TarantoolTupleImpl)) return false;
+        if (this == o) {
+            return true;
+        }
+        if (!(o instanceof TarantoolTupleImpl)) {
+            return false;
+        }
         TarantoolTupleImpl that = (TarantoolTupleImpl) o;
         return Objects.equals(getFields(), that.getFields());
     }

--- a/src/main/java/io/tarantool/driver/core/TarantoolConnectionSelectionStrategies.java
+++ b/src/main/java/io/tarantool/driver/core/TarantoolConnectionSelectionStrategies.java
@@ -37,28 +37,26 @@ public final class TarantoolConnectionSelectionStrategies {
 
         private TarantoolConnectionIterator connectionIterator;
         private final int maxAttempts;
+        private final Collection<TarantoolConnection> connections;
 
         RoundRobinStrategy(Collection<TarantoolConnection> connections) {
             this.connectionIterator = new TarantoolConnectionIterator(connections);
-            maxAttempts = connections.size();
+            this.maxAttempts = connections.size();
+            this.connections = connections;
         }
 
         @Override
         public TarantoolConnection next() throws NoAvailableConnectionsException {
             if (connectionIterator.hasNext()) {
-                TarantoolConnection connection = connectionIterator.next();
                 int attempts = 0;
-                while (!connection.isConnected() && attempts++ < maxAttempts) {
-                    connection = connectionIterator.next();
+                while (attempts++ < maxAttempts) {
+                    TarantoolConnection connection = connectionIterator.next();
+                    if (connection.isConnected()) {
+                        return connection;
+                    }
                 }
-                if (connection.isConnected()) {
-                    return connection;
-                } else {
-                    throw new NoAvailableConnectionsException();
-                }
-            } else {
-                throw new NoAvailableConnectionsException();
             }
+            throw new NoAvailableConnectionsException();
         }
     }
 
@@ -100,25 +98,22 @@ public final class TarantoolConnectionSelectionStrategies {
                     .collect(Collectors.groupingBy(
                             conn -> currentSize.getAndIncrement() / groupSize)).values().stream()
                     .map(TarantoolConnectionIterator::new)
+                    .filter(TarantoolConnectionIterator::hasNext)
                     .collect(Collectors.toList());
         }
 
         @Override
         public TarantoolConnection next() throws NoAvailableConnectionsException {
             if (iteratorsIterator.hasNext()) {
-                TarantoolConnection connection = iteratorsIterator.next().next();
                 int attempts = 0;
-                while (!connection.isConnected() && attempts++ < maxAttempts) {
-                    connection = iteratorsIterator.next().next();
+                while (attempts++ < maxAttempts) {
+                    TarantoolConnection connection = iteratorsIterator.next().next();
+                    if (connection.isConnected()) {
+                        return connection;
+                    }
                 }
-                if (connection.isConnected()) {
-                    return connection;
-                } else {
-                    throw new NoAvailableConnectionsException();
-                }
-            } else {
-                throw new NoAvailableConnectionsException();
             }
+            throw new NoAvailableConnectionsException();
         }
     }
 }

--- a/src/main/java/io/tarantool/driver/metadata/TarantoolSpaceMetadata.java
+++ b/src/main/java/io/tarantool/driver/metadata/TarantoolSpaceMetadata.java
@@ -115,6 +115,17 @@ public class TarantoolSpaceMetadata {
         return getFieldByPosition(fieldPosition).map(TarantoolFieldMetadata::getFieldName);
     }
 
+    @Override
+    public String toString() {
+        return "TarantoolSpaceMetadata{" +
+                "spaceId=" + spaceId +
+                ", ownerId=" + ownerId +
+                ", spaceName='" + spaceName + '\'' +
+                ", spaceFormatMetadata=" + spaceFormatMetadata +
+                ", spaceFormatMetadataAsList=" + spaceFormatMetadataAsList +
+                '}';
+    }
+
     /*
       - [281, 1, '_vspace', 'sysview', 0, {}, [{'name': 'id', 'type': 'unsigned'}, {'name': 'owner',
         'type': 'unsigned'}, {'name': 'name', 'type': 'string'}, {'name': 'engine',

--- a/src/test/java/io/tarantool/driver/core/CustomConnection.java
+++ b/src/test/java/io/tarantool/driver/core/CustomConnection.java
@@ -68,4 +68,14 @@ final class CustomConnection implements TarantoolConnection {
     @Override
     public void close() throws Exception {
     }
+
+    @Override
+    public String toString() {
+        return "CustomConnection{" +
+                "host='" + host + '\'' +
+                ", port=" + port +
+                ", count=" + count +
+                ", connected=" + connected +
+                '}';
+    }
 }

--- a/src/test/java/io/tarantool/driver/core/ParallelRoundRobinStrategyTest.java
+++ b/src/test/java/io/tarantool/driver/core/ParallelRoundRobinStrategyTest.java
@@ -107,7 +107,7 @@ public class ParallelRoundRobinStrategyTest {
         List<TarantoolConnection> connections = IntStream.range(1, 11)
                 .mapToObj(i -> new CustomConnection(String.format("127.0.0.%d", i), 3000 + i))
                 .peek(c -> {
-                    if (c.getPort() % 2 == 0) {
+                    if (c.getPort() < 3006) {
                         c.setConnected(false);
                     }
                 })
@@ -128,7 +128,7 @@ public class ParallelRoundRobinStrategyTest {
         futures.forEach(CompletableFuture::join);
 
         for (TarantoolConnection c : connections) {
-            if (((CustomConnection) c).getPort() % 2 == 0) {
+            if (((CustomConnection) c).getPort() < 3006) {
                 assertEquals(0, ((CustomConnection) c).getCount(), "Closed connection must have been used 0 times");
             } else {
                 assertEquals(20, ((CustomConnection) c).getCount(), "Active connection must have been used 20 times");

--- a/src/test/java/io/tarantool/driver/core/RoundRobinStrategyTest.java
+++ b/src/test/java/io/tarantool/driver/core/RoundRobinStrategyTest.java
@@ -101,7 +101,7 @@ public class RoundRobinStrategyTest {
         List<TarantoolConnection> connections = IntStream.range(1, 11)
                 .mapToObj(i -> new CustomConnection(String.format("127.0.0.%d", i), 3000 + i))
                 .peek(c -> {
-                    if (c.getPort() % 2 == 0) {
+                    if (c.getPort() < 3006) {
                         c.setConnected(false);
                     }
                 })
@@ -122,7 +122,7 @@ public class RoundRobinStrategyTest {
         futures.forEach(CompletableFuture::join);
 
         for (TarantoolConnection c : connections) {
-            if (((CustomConnection) c).getPort() % 2 == 0) {
+            if (((CustomConnection) c).getPort() < 3006) {
                 assertEquals(0, ((CustomConnection) c).getCount(), "Closed connection must have been used 0 times");
             } else {
                 assertEquals(20, ((CustomConnection) c).getCount(), "Active connection must have been used 20 times");

